clang/lib/Sema/SemaTemplate.cpp

```diff
@@ -3158,6 +3161,33 @@ checkBuiltinTemplateIdType(Sema &SemaRef, BuiltinTemplateDecl *BTD,
     int64_t N = Index.getExtValue();
     return Ts.getPackAsArray()[N].getAsType();
   }
+  case BTK__type_pack_dedup: {
+    assert(Converted.size() == 2 && "__builtin_type_pack_dedup should be given "
+                                    "a template and a parameter pack");
+    TemplateArgument Template = Converted[0];
+    TemplateArgument Ts = Converted[1];
+    if (Template.isDependent() || Ts.isDependent())
```

Comment:
I think the only reason we would need to hold off on building the template specialization would be if Ts contains an unexpanded pack, otherwise we are good to go, subsequent substitutions will produce the correct type list anyway.

This would also reduce the size of the argument list.

Be aware that holding this specialization will make this builtin appear in mangling, so I'd avoid doing that unless necessary.

Reply:
We could remove the dependency on the template, but I don't think we can ignore the dependency inside the types themselves.
Consider
```cpp
template <class T, class U, template <class...> class TL = TypeList>
struct Foo {
  // 1. the substitution has to be postponed until T and U are known.
  using result = __dedup<TypeList, T, U>;
  // 2. this is already computed right away.
  using result2 = __dedup<TypeList, int, int>; 
  // 3. this **could** be computed right away, but it is not now.
  using result3 = __dedup<TL, int, int>
};

static_assert(__is_same(TypeList<int>, Foo<int, int>::result));
static_assert(__is_same(TypeList<int, doubel>, Foo<int, double>::result));
```

- Is there something I'm missing that would allow (1) to be substituted earlier?
- Do you feel strongly that we need to do (3) ?

> Be aware that holding this specialization will make this builtin appear in mangling, so I'd avoid doing that unless necessary.

I do not think we can avoid this completely (see above) and it's not unprecedented: other builtin templates can already appear in mangling too when the arguments are dependent. So I do think it's necessary and I'm not sure which problems it carries.

Reply:
You are right that we can't drop out the builtin template specialization just yet if any of the arguments are dependent.

But we can still unique the list in that case. This actually even applies for unexpanded packs as well.

One advantage I see is that you would avoid the possibility of the user introducing multiple overloads that are equivalent, which you wouldn't be able to partial order later.

Ie

```C++
template <class...> struct A;
template <class T> void f(__builtin_type_pack_dedup<A, T>) {}
template <class T> void f(__builtin_type_pack_dedup<A, T, T>) {}
```
This would be ill-formed because `f` is being redefined.

This also increases the ABI stability for the same reason. If the user introduces another repeated template argument, it's still results in the same templated entity with the same mangling.

> Do you feel strongly that we need to do (3) ?

See above.

> I do not think we can avoid this completely (see above) and it's not unprecedented: other builtin templates can already appear in mangling too when the arguments are dependent. So I do think it's necessary and I'm not sure which problems it carries.

Right, it's not completely unavoidable, but at least for the dependent template name case that we can resolve early, it makes it so this builtin can be introduced without causing an ABI break.